notebooks/Batchelor.ipynb

```diff
@@ -0,0 +1,348 @@
+    "    G = -ufl.div(v_t)*trial_p*ufl.dx\n",
+    "    D = -p_t*ufl.div(trial_v)*ufl.dx\n",
+    "    S = K + G + D\n",
+    "\n",
+    "    zeroes = Constant(domain, default_scalar_type(0.0))\n",
+    "    L = zeroes*p_t*ufl.dx\n",
+    "\n",
+    "    # assemble problem solver and solve\n",
+    "    problem = LinearProblem(S, L, bcs=[bc1, bc2, bc3, bc4], petsc_options={\"ksp_type\": \"preonly\", \"pc_type\": \"lu\", \"pc_factor_mat_solver_type\": \"mumps\"})\n",
+    "    uh = problem.solve()\n",
+    "\n",
+    "    vh = uh.sub(0).collapse()\n",
+    "\n",
+    "    return vh"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "de15d3f6-6e1d-41d3-8f56-bcf76a7b1a59",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def error(vh):\n",
+    "    x = ufl.SpatialCoordinate(vh.function_space.mesh)\n",
+    "    vex = u_exact(x)\n",
+    "    error_L2 = assemble_scalar(form((vh - vex)**2 * ufl.dx))\n",
+    "    error_L2 = np.sqrt(MPI.COMM_WORLD.allreduce(error_L2, op=MPI.SUM))\n",
+    "\n",
+    "    xs = np.linspace(0, 1, 1001)  # third argument is arbitrary\n",
+    "    xgrid, ygrid = np.meshgrid(xs,xs)\n",
+    "    xyzs = np.stack((xgrid.flatten(), ygrid.flatten(), np.zeros_like(xgrid.flatten())), axis=1)\n",
+    "    cells = get_first_cells(xyzs, vh.function_space.mesh)\n",
+    "    vh_vals = vh.eval(xyzs, cells)[:,0]\n",
+    "\n",
+    "    vex_vals = u_exact_np(xyzs.transpose())\n",
+    "    error_max = np.max(np.abs(vh_vals - vex_vals))\n",
+    "    error_max = MPI.COMM_WORLD.allreduce(error_max, op=MPI.MAX)\n",
+    "    \n",
+    "    return error_L2, error_max"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a035fd02-45c8-4a9b-8bf1-52f0e300aafc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# define a utility function to help with visualization\n",
+    "def get_first_cells(x, mesh):\n",
+    "    \"\"\"\n",
+    "    Given a list of points and a mesh, return the first cells that each point lies in in the mesh.\n",
+    "\n",
+    "    Arguments:\n",
+    "      * x    - coordinates of points\n",
+    "      * mesh - mesh\n",
+    "\n",
+    "    Returns:\n",
+    "      * first_cells - a list of cells corresponding to each point in x\n",
+    "    \"\"\"\n",
+    "    tree = geometry.bb_tree(mesh, mesh.geometry.dim)\n",
+    "    xl = x\n",
+    "    if len(x.shape)==1: xl = [x]\n",
+    "    first_cells = []\n",
+    "    for x0 in xl:\n",
+    "        cell_candidates = geometry.compute_collisions_points(tree, x0)\n",
+    "        cell = geometry.compute_colliding_cells(mesh, cell_candidates, x0).array\n",
+    "        assert len(cell) > 0\n",
+    "        first_cells.append(cell[0])\n",
+    "    return first_cells"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "acfef0b4-bd65-4f4f-b59e-1b6e85fea0ba",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def visualize(vh):\n",
```

Comment:
This repository has a `utils` module that contains various plotting functions.  Let's talk about replacing your hand-rolled version with that so you can get rid of this cell and call `utils.plot_vector_glyphs` instead.